### ISSUE TITLE
Support compile-time weaving and dependency injection [SPR-16359]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-framework#20906]
**[Remo](https://jira.spring.io/secure/ViewProfile.jspa?name=remmeier)** opened **[SPR-16359](https://jira.spring.io/browse/SPR-16359?redirect=false)** and commented

Spring incurs high computational costs during startup to wire beans together and perform AOP. With newer technologies like docker or serverless/faas architectures there is a trend to move more and more work into the compile-step where possible. On the one side it allows to catch issues early. On the other side it allows to perform more optimizations like

- tree-shaking (with something like proguard) to remove unnecessary parts
- AoT compilation with Java 9+
- Jlink with Java9+
- and in general do the wiring only once at compile-time instead of each time at runtime.
- exclude AutoConfigurations not necessary for a particular application

having such a thing in place would drastically improve startup performance, but can also would make the application more lightweight, saving space and memory.

Other projects in the area is, for example, https://github.com/google/dagger.

Doing something in this area seems in particular important for serverless/faas architectures (https://github.com/spring-cloud/spring-cloud-function/issues/132). Without a small footprint and in particular very fast cold start performance, Spring will not be applicable in such areas.

---

**Affects:** 5.0.2

3 votes, 6 watchers

**Comment by remmeier** [comment]
maybe more followups in the area:

- micronaut now basically did what you are looking for: https://github.com/micronaut-projects/micronaut-spring, at lest a good start
- https://devclass.com/2019/03/08/red-hat-brews-up-quarkus-to-pour-java-into-smaller-containers/ does some work with native code & Quark

still think theire is tremendous in the area or also the danger for getting obsolete by competitors. we ran quite large number of spring boot micro services, but the footprint and startup costs are too large compared to experiments with other frameworks.

**Comment by dsyer** [comment]
Please can you share some code with the comparisons and analysis that you did? We publish benchmarks on Spring, and are finding lots of optimizations. The other frameworks might be as amazing as they claim but I haven't seen any evidence yet that I could reproduce.

**Comment by davidkiss** [comment]
See here for performance benchmark results: https://www.techempower.com/benchmarks/ and source code for both the micronaut and spring code used: https://github.com/TechEmpower/FrameworkBenchmarks/tree/master/frameworks/Java.

**Comment by dsyer** [comment]
That's a data point I guess. But the techempower benchmarks don't really compare "footprint" and "startup costs", which is what is claimed as a benefit of compile-time weaving. Sadly, these benchmarks are also not perceived as a good comparison of real application performance. The specs for the individual apps are not really representative of real application workloads.

The evidence that we have is pointing to likely marginal gains from changing to a compile time approach, and we have already improved some metrics in Spring Boot 2.2/Spring 5.2 by some large factors (e.g. some apps that start 15x faster, and apps that start in <10MB heap). This issue is still open, and we are continuing to discuss the topic with open minds, but personally I'm not convinced it's going to be worth it.

**Comment by remmeier** [comment]
"The evidence that we have is pointing to likely marginal gains from changing to a compile time approach"

If it is just about replacing some reflection calls, then yes true. Reflection can be rather fast. spring-aop is an area where I would see benefits. Some performance issues on startup also come from popular third-party like Hibernate. But still real and main benefit for opening up this ticket was also to provide an entry point for further optimizations as mentioned in the description. compile-time injection would play well together with tree shaking and aot/graal. graal already brings benefits on its own. tree shaking is rather hard/cumbercomse to achieve for a spring boot application due to all the reflection/aop/discovery. Would be amazed if there are not some substantial results when bringing all three together. In other languages/frameworks like Angular does benefits are well established.

**Comment by cmyker** [comment]
Just wanted to highlight one area where functionality mentioned above get increasingly important - serverless. Even a milisecond delay of a cold start time is important there, since it piles up into a big number... 
We can certainly discuss that not only spring has issues there, but absence of JIT would be a performance penalty... But it is matter of fact that trend is moving there, and most people are willing to neglect JIT issues in favour of saving costs by converting certain services to responding on demand. 
This is getting increasingly important since more and more organisations split there infrastructure into smaller domains served by certain microservices that might not be feasible to run all the time or setup complex autoscaling functionality.

**Comment by jhoeller** [comment]
This ticket has generally been superseded by Spring AOT (which performs a lot of scanning and introspection at build time and transforms common application configuration accordingly), along with the GraalVM Native Image support in Spring Framework 6. And for the startup time problem, there is also CRaC and AppCDS now in Spring Framework 6.1.

For concrete improvement requests around our AOT efforts, please create dedicated tickets for it.
